Advanced Code/FileSystemVisitor.cs

```diff
@@ -1,54 +1,209 @@
 ﻿using System;
 using System.IO;
+using System.Collections.Generic;
 
 namespace Advanced_Code
 {
-	public class FileSystemVisitor
-	{
+    public class FileSystemVisitor
+    {
         private readonly string _folderPath;
+        private bool _abortSearch;
+        private List<FileSystemInfo> _excludedFiles;
+        private List<DirectoryInfo> _excludedFolders;
 
-        // we can use this:
-        // private readonly Func<FileSystemInfo, bool> _filter;
+        private readonly FileFilterDelegate _fileFilter;
+        private readonly SubFolderFilterDelegate _subFolderFilter;
+        private List<FileSystemInfo> _foundFiles = new List<FileSystemInfo>(); 
 
-        private readonly FileSystemFilterDelegate _filter;
+        public event EventHandler Start;
+        public event EventHandler Finish;
+        public event EventHandler Abort;
+        public event EventHandler<FileSystemInfo> FileFound;
+        public event EventHandler<FileSystemInfo> DirectoryFound;
+        public event EventHandler<List<FileSystemInfo>> FilteredFilesFound;
+        public event EventHandler<FileSystemInfo> FilteredDirectoryFound;
+        public event EventHandler<UserPromptEventArgs> UserPrompt;
 
 
-        public FileSystemVisitor(string rootPath, FileSystemFilterDelegate filter)
+        public FileSystemVisitor(string rootPath, FileFilterDelegate filter, SubFolderFilterDelegate subFolderFilter)
         {
             _folderPath = rootPath;
-            _filter = filter ?? (item => true); 
+            _fileFilter = filter ?? (item => true);
+            _subFolderFilter = subFolderFilter ?? (item => true);
+            _excludedFiles = new List<FileSystemInfo>();
+            _excludedFolders = new List<DirectoryInfo>();
         }
 
+        public FileSystemVisitor(string rootPath, FileFilterDelegate filter)
+       : this(rootPath, filter, null)
+        {
+        }
+
+        public FileSystemVisitor(string rootPath, SubFolderFilterDelegate subFolderFilter)
+       : this(rootPath, null, subFolderFilter)
+        {
+        }
+
+
         public FileSystemVisitor(string rootPath)
-        : this(rootPath, null)
+        : this(rootPath, null, null)
         {
         }
 
         public IEnumerable<FileSystemInfo> Inspect()
         {
-            return InspectDirectory(new DirectoryInfo(_folderPath));
+            var list = GetListOfFilesfromDirectory(new DirectoryInfo(_folderPath));
+            if (list == null) Console.WriteLine("There is no such directory, no more");
```

Comment:
This condition will never be fulfilled as GetListOfFilesfromDirectory cannot return null